tests/kernel/semaphore/sys_sem/src/main.c

```diff
@@ -0,0 +1,212 @@
+
+	k_sem_give(&usage_sem);
+	k_sem_take(&sync_sem, K_FOREVER);
+	zassert_equal(flag, LOW_PRIO, "low prio value error");
+
+	k_thread_join(&tdata2, K_FOREVER);
+	k_thread_join(&tdata3, K_FOREVER);
+	k_thread_join(&tdata4, K_FOREVER);
+
+	k_thread_create(&tdata1, stack1, STACK_SIZE,
+			sem_thread_give_uninit, NULL, NULL,
+			NULL, PRIO, K_USER | K_INHERIT_PERMS,
+			K_NO_WAIT);
+	k_sleep(K_MSEC(20));
+	k_thread_join(&tdata1, K_FOREVER);
+}
+
+static void multi_thread_sem_give(void *p1, void *p2, void *p3)
+{
+	count++;
+	k_sem_give(&limit_sem);
+
+	if (count < TOTAL_MAX)
+		zassert_equal(k_sem_count_get(&limit_sem), count, "multi get sem error");
+	else
+		zassert_equal(k_sem_count_get(&limit_sem), SEM_MAX_VAL, "count > SEM_MAX_VAL");
+
+	k_sem_take(&sync_sem, K_FOREVER);
+}
+
+static void multi_thread_sem_take(void *p1, void *p2, void *p3)
+{
+	k_sem_take(&limit_sem, K_FOREVER);
+	count--;
+
+	if (count >= 0)
+		zassert_equal(k_sem_count_get(&limit_sem), count, "multi take sem error");
+	else
+		zassert_equal(k_sem_count_get(&limit_sem), 0, "count < SEM_INIT_VAL");
+
+	k_sem_give(&sync_sem);
+}
+
+/**
+ * @brief Test max semaphore can be give and take with multiple thread
+ *
+ * @details
+ * - Define and initialize semaphore and thread.
+ * - Give sem by multiple threads.
+ * - Verify more than max count about semaphore can reach.
+ * - Take sem by multiple threads and verify if sem count is correct.
+ *
+ * @ingroup kernel_sys_sem_tests
+ */
+void test_multi_thread_sem_limit(void)
+{
+	k_sem_init(&limit_sem, SEM_INIT_VAL, SEM_MAX_VAL);
+	k_sem_init(&sync_sem, SEM_INIT_VAL, SEM_MAX_VAL);
+
+	count = 0;
+	for (int i = 1; i <= TOTAL_MAX; i++) {
+		k_thread_create(&multi_tid_give[i], multi_stack_give[i], STACK_SIZE,
+				multi_thread_sem_give, NULL, NULL, NULL,
+				i, K_USER | K_INHERIT_PERMS, K_NO_WAIT);
+	}
+
+	k_sleep(K_MSEC(50));
+
+	count = SEM_MAX_VAL;
+	for (int i = 1; i <= TOTAL_MAX; i++) {
+		k_thread_create(&multi_tid_take[i], multi_stack_take[i], STACK_SIZE,
+				multi_thread_sem_take, NULL, NULL, NULL,
+				PRIO, K_USER | K_INHERIT_PERMS, K_NO_WAIT);
+	}
+}
+
+void test_main(void)
+{
+	k_thread_access_grant(k_current_get(), &usage_sem, &sync_sem,
+			      &limit_sem, &tdata1, &tdata2, &tdata3,
+			      &multi_tid_give[0], &multi_tid_give[1],
+			      &multi_tid_give[2], &multi_tid_give[3],
+			      &multi_tid_give[4], &tdata4, &stack1,
+			      &stack2, &stack3, &stack4,
+			      &multi_stack_take[0], &multi_stack_take[1],
+			      &multi_stack_take[3], &multi_stack_take[4],
+			      &multi_stack_take[2], &multi_stack_give[0],
+			      &multi_stack_give[1], &multi_stack_give[2],
+			      &multi_stack_give[3], &multi_stack_give[4],
+			      &multi_tid_take[0], &multi_tid_take[1],
+			      &multi_tid_take[2], &multi_tid_take[3],
+			      &multi_tid_take[4]);
+
+	ztest_test_suite(test_kernel_sys_sem,
+			 ztest_user_unit_test(test_multiple_thread_sem_usage),
+			 ztest_user_unit_test(test_multi_thread_sem_limit));
```

Comment:
Do we really need a completely new test just to add two cases?  Maybe fold this into an existing test, or does it not fit anywhere?

Reply:
Be honest, those cases what I added is integration tests, the existing tests we recorded to module tests. And we want has two reports(module and integration) at last, so it is easy to produce documents if created a new directory.
Do you thinks we need move those to existing tests?